### PR TITLE
GROK-13817: Chem: Highlighting multiple substructures

### PR DESCRIPTION
**Pull request by MariaDolotova**

### CODE REVIEW

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 1]
```diff
@@ -6,13 +6,26 @@ import * as DG from 'datagrok-api/dg';
...
-  substruct: any;
+  substruct: ISubstruct;
   molString: string;
-  scaffoldMolString: string;
+  scaffoldMolString: IColoredScaffold[];
+}
+
+export interface IColoredScaffold {
+  molString: string,
+  color?: number[]
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:28]: Why is it number[]?

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 2]
```diff
@@ -6,13 +6,26 @@ import * as DG from 'datagrok-api/dg';
...
+}
 
 interface IMolInfo {
   //mol: RDMol | null; // null when molString is invalid?
   molCtx: IMolContext;
-  substruct: any;
+  substruct: ISubstruct;
   molString: string;
-  scaffoldMolString: string;
+  scaffoldMolString: IColoredScaffold[];
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:23]: The naming makes no sense. Why a "string" is a collection of scaffolds?

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 3]
```diff
@@ -6,13 +6,26 @@ import * as DG from 'datagrok-api/dg';
...
+  bonds?: number[],
+  highlightAtomColors?: {[key: number]: number[]},
+  highlightBondColors?: {[key: number]: number[]}
+}
 
 interface IMolInfo {
   //mol: RDMol | null; // null when molString is invalid?
   molCtx: IMolContext;
-  substruct: any;
+  substruct: ISubstruct;
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:21]: Why does molInfo contain all of that? Perhaps a class should be renamed?

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 4]
```diff
@@ -167,17 +199,48 @@ M  END
     };
   }
 
-  _fetchMol(molString: string, scaffoldMolString: string, molRegenerateCoords: boolean,
-    scaffoldRegenerateCoords: boolean, details: object = {}): IMolInfo {
-    const name = molString + ' || ' + scaffoldMolString + ' || ' +
+  _addAtomsOrBonds(fromAtomsOrBonds: number[], toAtomsOrBonds: number[]) {
+    fromAtomsOrBonds?.forEach((it) => {
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:203]: Prefer the usual for loop

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 5]
```diff
@@ -167,17 +199,48 @@ M  END
...
+      if (!toAtomsOrBonds?.includes(it))
+        toAtomsOrBonds?.push(it);
+    });
+  }
+
+  _addColorsToBondsAndAtoms(mainSubstr: ISubstruct, color?: number[], tempSubstr?: ISubstruct){
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:209]: Follow our coding notation. Spaces, annotated return types

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 6]
```diff
@@ -242,14 +305,31 @@ M  END
     }
   }
 
-  _initScaffoldString(colTemp: any, tagName: string) {
+  _initScaffoldString(colTemp: any, tagName: string): IColoredScaffold[] {
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:308]: Annotate return types

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts [hunk 7]
```diff
@@ -242,14 +305,31 @@ M  END
...
     if (scaffoldString?.endsWith(this.WHITE_MOLBLOCK_SUFFIX)) {
-      scaffoldString = null;
       if (colTemp[tagName])
         delete colTemp[tagName];
+      return [];
+    }
+    return scaffoldString ? [{molString: scaffoldString}] : [];
+  }
+
+  _initScaffoldArray(colTemp: any, tagName: string): IColoredScaffold[] {
```
Comment (skalkin) [inline packages/Chem/src/rendering/rdkit-cell-renderer.ts:318]: Annotate return types

**Review by skalkin — CHANGES_REQUESTED** [review]

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts (hunk 3 above)
Comment (MariaDolotova) [reply packages/Chem/src/rendering/rdkit-cell-renderer.ts:21]: Yes, I renamed it to IMolRenderingInfo (it contains information required to render molecule with highlighted scaffols)

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts (hunk 2 above)
Comment (MariaDolotova) [reply packages/Chem/src/rendering/rdkit-cell-renderer.ts:23]: Renamed

File: packages/Chem/src/rendering/rdkit-cell-renderer.ts (hunk 1 above)
Comment (MariaDolotova) [reply packages/Chem/src/rendering/rdkit-cell-renderer.ts:28]: RDKit takes a color as an array of floats (so its is an RGB% format)

**Pull request merged**